pkg/services/oracle/oracle.go

```diff
@@ -159,20 +169,39 @@ func NewOracle(cfg Config) (*Oracle, error) {
 		return nil, errors.New("no wallet account could be unlocked")
 	}
 
-	if o.Client == nil {
-		var client http.Client
-		client.Transport = &http.Transport{DisableKeepAlives: true}
-		client.Timeout = o.MainCfg.RequestTimeout
-		o.Client = &client
-	}
 	if o.ResponseHandler == nil {
 		o.ResponseHandler = defaultResponseHandler{}
 	}
 	if o.OnTransaction == nil {
 		o.OnTransaction = func(*transaction.Transaction) error { return nil }
 	}
-	if o.URIValidator == nil {
-		o.URIValidator = defaultURIValidator
+	if o.URIResolver == nil {
+		o.URIResolver = defaultURIResolver
+	}
+	if o.Client == nil {
+		var client http.Client
+		client.Transport = &http.Transport{
+			DisableKeepAlives: true,
+			DialTLSContext: func(ctx context.Context, network, address string) (net.Conn, error) {
+				if !o.MainCfg.AllowPrivateHost {
+					ip, err := o.URIResolver(network, address)
+					if err != nil {
+						return nil, fmt.Errorf("%w: address %s failed validation: %s", oracleRedirectionErr, address, err)
+					}
+					network = ip.Network()
+					address = ip.IP.String() // TODO: without zone?
+				}
+				return net.Dial(network, address)
```

Comment:
> The returned net.Conn is assumed to already be past the TLS handshake.

Reply:
I've checked, we can safely customize only `DialContext`, see the updated comment.

Reply:
Yeah, I've seen that and it simplifies things a lot for us.